src/sinol_make/commands/run/__init__.py

```diff
@@ -201,37 +203,72 @@ def execute(self, execution):
-		if code == 35072:
-			result["Status"] = "TL"
-		elif "Status" not in result.keys():
-			result["Status"] = "RE"
-		elif result["Status"] == "OK":
-			if os.system("diff -q -Z %s %s >/dev/null"
-						% (output_file, self.get_output_file(test))):
-				result["Status"] = "WA"
-			elif result["Time"] > time_limit:
+		if self.args.time_tool == 'oiejq':
+			command = "MEM_LIMIT=%sK MEASURE_MEM=true timeout -k %ds -s SIGKILL %ds %s %s <%s >%s 2>%s" \
+					% (math.ceil(memory_limit), hard_time_limit_in_s,
+						hard_time_limit_in_s, timetool_path,
+						program, test, output_file, result_file)
+			code = os.system(command)
+			result = {}
+			with open(result_file) as r:
+				for line in r:
+					line = line.strip()
+					if ": " in line:
+						(key, value) = line.split(": ")[:2]
+						result[key] = value
+			if "Time" in result.keys():
+				result["Time"] = self.parse_time(result["Time"])
+			if "Memory" in result.keys():
+				result["Memory"] = self.parse_memory(result["Memory"])
+			if code == 35072:
 				result["Status"] = "TL"
-			elif result["Memory"] > memory_limit:
-				result["Status"] = "ML"
-		else:
-			result["Status"] = result["Status"][:2]
-		return result
+			elif "Status" not in result.keys():
+				result["Status"] = "RE"
+			elif result["Status"] == "OK":
+				if os.system("diff -q -Z %s %s >/dev/null"
+							% (output_file, self.get_output_file(test))):
+					result["Status"] = "WA"
+				elif result["Time"] > time_limit:
+					result["Status"] = "TL"
+				elif result["Memory"] > memory_limit:
+					result["Status"] = "ML"
+			else:
+				result["Status"] = result["Status"][:2]
+			return result
+		elif self.args.time_tool == 'time':
+			if sys.platform == 'darwin':
+				command = 'launchctl limit memlock %s; gtimeout -k %ds %ds gtime -f "%%U\\n%%M\\n%%x" -o %s %s <%s >%s' \
+					% (math.ceil(memory_limit) * 1024, hard_time_limit_in_s,
+						hard_time_limit_in_s, result_file, program, test, output_file)
+			elif sys.platform == 'linux':
+				command = 'ulimit -v %s; timeout -k %ds %ds time -f "%%U\\n%%M\\n%%x" -o %s %s <%s >%s' \
+					% (math.ceil(memory_limit), hard_time_limit_in_s,
+						hard_time_limit_in_s, result_file, program, test, output_file)
+				
+			code = os.system(command)
+			result = {}
+			lines = open(result_file).readlines()
+			if len(lines) == 3:
+				result["Time"] = round(float(lines[0].strip()) * 1000)
+				result["Memory"] = int(lines[1].strip())
+				if lines[2].strip() == "0":
+					result["Status"] = "OK"
+				else:
+					result["Status"] = "RE"
+
+			if code != 0:
+				result["Status"] = "TL"
+			if result["Status"] == "OK":
+				if os.system("diff -q -Z %s %s >/dev/null"
+							% (output_file, self.get_output_file(test))):
+					result["Status"] = "WA"
+				elif result["Time"] > time_limit:
+					result["Status"] = "TL"
+				elif result["Memory"] > memory_limit:
+					result["Status"] = "ML"
+			return result
+		elif self.args.time_tool == 'python':
+			pass
```

Comment:
The code becomes quite lengthy and they are implementation details, so it might be better to split the existing code into smaller functions.

Reply:
It should be better now 80b92ca6682a772342abb132bea23d0a09d3cb92